README.md
## VideoAudioRecorderJs

Single Library for Client Side Video & Audio Recording

Older Version of this Project was supporting video audio recording mechanisam implemented using whammy.js and Recorder.js but with the 1.0.0 version release
We have used [MediaStream Recording](https://www.w3.org/TR/mediastream-recording/) Spec as the main recroder but legacy browsers which does not support for this new spec will fallback to old recroder implementation, below are the browsers that are currently supported by the script

* Chrome
* Firefox
* Opera

Note : Safari is not supported

### Usage
* NPM : `npm install videorecorderjs --save`
* Bower : `bower install videorecorderjs --save`

Or linking script files

```html
<script src="../dist/VideoRecorderJS.min.js" type="text/javascript"></script>
```

Initializing 

```html

    var virec = new VideoRecorderJS.init(
            {
                resize: 0.8,
                webpquality: 0.9,
                framerate: 15,
                videotagid: "viredemovideoele",
                videoWidth: "640",
                videoHeight: "480",
                log: true,
                workerPath : "../dist/recorderWorker.js"
            },
            function () {
                //success callback. this will fire if browsers supports
            },
            function (err) {
                //onerror callback, this will fire for mediaErrors
                if (err.name == "BROWSER_NOT_SUPPORTED") {
                    //handler code goes here
                } else if (err.name == "PermissionDeniedError") {
                    //handler code goes here
                } else if (err.name == "NotFoundError") {
                    //handler code goes here
                } else {
                    throw 'Unidentified Error.....';
                }

            }
    );

```
* `resize` : indicates the recorded video dimentions relative to the actual size
* `webpquality`     : this indicate the quality of a single frame in the video
* `framerate`       : frame rate


### Here is a Complete Example
check the `demo` folder for working example

```html

<html>
<head>
</head>
<body>
<div id="videorecorder">
    <video id="viredemovideoele"></video>
    <span style="font-size:20px;" id="countdown"></span>
</div>

<input id="playback" value="PlayBack" type="button"/>
<input id="clearrecording" value="Clear Recording" type="button"/>
<input id="startRecrodBut1" value="Start Recording" type="button"/>
<input id="stopRecBut1" value="Stop Recording" type="button"/>
<input id="uploadrecord" value="Upload Recording" type="button"/>


</br>
<p id="status"></p>
<video id="recordedvideo" controls></video>
<audio id="audiored" controls></audio>
<a id="downloadurl">Download</a>
<div id="progressNumber" style="font-size:20px;"></div>




<script src="../dist/VideoRecorderJS.min.js" type="text/javascript"></script>

<script type="text/javascript">


    var startRecord = document.getElementById("startRecrodBut1");
    var stopRecord = document.getElementById("stopRecBut1");
    var countdownElement = document.getElementById("countdown");
    var playBackRecord = document.getElementById("playback");
    var discardRecordng = document.getElementById("clearrecording");
    var uploadrecording = document.getElementById("uploadrecord");
    var progressNumber = document.getElementById("progressNumber");


    var virec = new VideoRecorderJS.init(
            {
                resize: 0.8, // recorded video dimentions are 0.4 times smaller than the original
                webpquality: 0.5, // chrome and opera support webp imags, this is about the aulity of a frame
                framerate: 15,  // recording frame rate
                videotagid: "viredemovideoele",
                videoWidth: "640",
                videoHeight: "480",
                log: true,
                mediaRecorderType : "webscript",
                workerPath : "../dist/recorderWorker.js"
            },
            function () {
                //success callback. this will fire if browsers supports
            },
            function (err) {
                //onerror callback, this will fire for mediaErrors
                if (err.name == "BROWSER_NOT_SUPPORTED") {
                    //handler code goes here
                } else if (err.name == "PermissionDeniedError") {
                    //handler code goes here
                } else if (err.name == "NotFoundError") {
                    //handler code goes here
                } else {
                    throw 'Unidentified Error.....';
                }

            }
    );

    startRecord.addEventListener("click", function () {
        virec.startCapture(); // this will start recording video and the audio
        stopCountDown();
        startCountDown();
    });

    stopRecord.addEventListener("click", function () {
        /*
         stops the recording and after recording is finalized oncaptureFinish call back
         will occur
         */
        virec.stopCapture(oncaptureFinish);
        stopCountDown();
    });

    playBackRecord.addEventListener("click", function () {
        /*
         Clientside playback,
         */
        virec.play();
    });

    discardRecordng.addEventListener("click", function () {
        /*
         Clears the current recorded video + audio allowing
         another recording to happen
         */
        virec.clearRecording();
        stopCountDown();
    });

    uploadrecording.addEventListener("click", function () {
        /*
         Uploading the content to the server, here I have sliced the blobs into chunk size
         of 1048576 bits so that uploading time will reduce.
         Gmail uses this same technique when we attach some files to a mail, it slize the file
         in the client side and then uploads chunk by chunk
         */
        var uploadoptions = {
            blobchunksize: 1048576,
            requestUrl: "php/fileupload.php",
            requestParametername: "filename",
            videoname: "video.webm",
            audioname: "audio.wav"
        };
        virec.uploadData(uploadoptions, function (totalchunks, currentchunk) {
            /*
             This function will callback during, each successfull upload of a blob
             so you can use this to show a progress bar or something
             */
            progressNumber.innerHTML = ((currentchunk / totalchunks) * 100);
            console.log(currentchunk + " OF " + totalchunks);
        });
    });


    //------------------------------- few functions that demo, how to play with the api --------------------------

    var countdowntime = 15;
    var functioncalltime = 0;
    var timerInterval = null;

    function oncaptureFinish(result) {
        document.getElementById('status').innerHTML = "";
        result.forEach(function (item) {
            if (item.type == "video") {
                var videoblob = item.blob;
                var videobase64 = window.URL.createObjectURL(videoblob);
                document.getElementById('recordedvideo').src = videobase64;
                document.getElementById('downloadurl').style.display = '';
                document.getElementById('downloadurl').href = videobase64;
                document.getElementById('status').innerHTML = document.getElementById('status').innerHTML + "video=" + Math.ceil(videoblob.size / (1024)) + "KB";

            } else if (item.type == "audio") {
                var audioblob = item.blob;
                document.getElementById('audiored').src = window.URL.createObjectURL(audioblob);
                document.getElementById('status').innerHTML = document.getElementById('status').innerHTML + "Audio=" + Math.ceil(audioblob.size / (1024)) + "KB";
            }
        });
    }

    function setCountDownTime(time) {
        countdownElement.innerHTML = time;
        return time;
    }


    function startCountDown() {
        if (timerInterval == null) {
            functioncalltime = countdowntime;
            var value = setCountDownTime(functioncalltime);
            timerInterval = setInterval(function () {
                var value = setCountDownTime(--functioncalltime);
                if (value == 0) {
                    clearInterval(timerInterval);
                    virec.stopCapture(oncaptureFinish);
                }
            }, 1000);
        }
    }

    function stopCountDown() {
        if (timerInterval) {
            clearInterval(timerInterval);
            timerInterval = null;
        }
    }


</script>
</body>
</html>


```
### Server Side Code
I have used php but you can use any server side language


## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## Contributing
**Bug fixes** and **new features** can be proposed using [pull requests](https://github.com/imalhasaranga/VideoRecorderJs/pulls).
Please read the [contribution guidelines](CONTRIBUTION.md) before submitting a pull request.

## Credits

- [Imal Hasaranga Perera](https://github.com/imalhasaranga)
- [All Contributors](../../contributors)


## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.



CHANGELOG.md
# Changelog

All changes to `VideoAudioRecorderJs` will be documented in here

### Minor Bug Fixes
- demo bug fixed

## 1.0.1 - 2017-02-22

### Old Repo Updated with 1.0.0 Release
- This is the first release version of VideoAudioRecorderJs

## 1.0.0 - 2017-02-21


###Fixes, Changes

1. increased the flexibility of the api
2. provided support for Opera latest version
3. removed the dependancy of JQuery and make it pure javascript
4. uploading time reduced by uploading blob as chunks
5. old Library previously supported direct mp3 recording using a library provided by [Octavian Naicu](https://github.com/nusofthq/Recordmp3js) but with this version I removed it because of few bugs in that project, and it takes conciderable time to encode to mp3



im365.js



var VIRecorder = (function(){
    

    window.URL =    window.URL || 
                    window.webkitURL;
        
    navigator.getUserMedia  =   navigator.getUserMedia || 
                                navigator.webkitGetUserMedia || 
                                navigator.mozGetUserMedia || 
                                navigator.msGetUserMedia;

    window.AudioContext     =   window.AudioContext || 
                                window.webkitAudioContext;

    window.BlobBuilder      =   window.MozBlobBuilder || 
                                window.WebKitBlobBuilder || 
                                window.BlobBuilder;

    var canvas;
    var ctx;


    var recorder;
    var localStream;
    var audio_context;

    var UploadingURL = "";

    var videoAudioSync = null;
    var recrodinterval = null;

    var quality = 1;
    var framerate = 15;
    var frames = [];
    var webp_quality = 0.8;

    var audioElement;
    var videoElement;
    var audioBlobURL;
    var videoBlobURL;

    var countdowntime = 15;
    var countdownElement= null;
    var startTime = null;

    function initVIRecorder(options) {
        var startButton;
        var stopButton;
        var vw = "320";
        var vh = "240";
        
        
        var startRecrodBut = "startRecrodBut";
        var stopRecBut = "stopRecBut";
        var palaybackbut = "playback";
        var discardrecrd = "clearrecording";

       
       
        quality = evaluateQuality(options.initVIRecorder);
        if(options.startButtonId != undefined){   startRecrodBut = options.startButtonId+""  }
        if(options.stopButtonId != undefined) {   stopRecBut = options.stopButtonId+""       }   
        if(options.videoWidth != undefined)   {   vw = options.videoWidth+""                 }    
        if(options.videoHeight != undefined)  {   vh = options.videoHeight+""                } 
        if(options.uploadURL != undefined)    {   UploadingURL = options.uploadURL+""        }     
        if(options.webpquality != undefined)  {   webp_quality = options.webpquality         }     
        if(options.framerate != undefined)    {   framerate = options.framerate              }     
        if(options.countdowntime != undefined){   countdowntime = options.recordtime         }     


        audioElement = document.querySelector('audio'); 
        videoElement = document.getElementById('im385videorecele');   
        countdownElement = document.getElementById("countdown");
        var startRecord = document.getElementById(startRecrodBut);
        var stopRecord  = document.getElementById(stopRecBut);
        var playBackRecord = document.getElementById(palaybackbut);
        var discardRecordng  = document.getElementById(discardrecrd);

        
        startRecord.addEventListener("click" , function(){
             startCapture();
        });
        stopRecord.addEventListener("click" , function(){
             stopCapture();
        });
         playBackRecord.addEventListener("click" , function(){
             playBack();
        });
        discardRecordng.addEventListener("click" , function(){
            discardRecord();
        });


        prepareVideoElement(videoElement);
        setCountDownTime(countdowntime);
        canvas =  document.createElement('canvas');
        ctx = canvas.getContext('2d');
        

        try {
            audio_context = new AudioContext;
            navigator.getUserMedia(
            {
                audio: true, 
                video: true
            }, 
            function(stream){
                
                var input = audio_context.createMediaStreamSource(stream);
                localStream = window.URL.createObjectURL(stream);
                videoElement.src = localStream;
                
                var zeroGain = audio_context.createGain();
                zeroGain.gain.value = 0;
                input.connect(zeroGain);
                zeroGain.connect(audio_context.destination);
                recorder = new Recorder(input);
            }, 
            function(e) {
                alert('Audio Input is Not Found : ' + e);
            });

        } catch (e) {
            alert('your browser does not support');
        }


        function prepareVideoElement(videoelement){
            videoelement.width = vw;
            videoelement.height = vh;
            videoelement.autoplay = true;
            videoelement.muted = true;
        }

    };



    function startCapture() {
        startCountDown(null);
        startTime = new Date().getTime();
        // ------- Video Recording started ---------------------------------
        var newWidth = canvas.width  = parseInt(quality*videoElement.clientWidth);
        var newHeight = canvas.height = parseInt(quality*videoElement.clientHeight);
        var timmer  = parseInt(1000 /framerate);

        recrodinterval = setInterval(function(){
            ctx.drawImage(videoElement, 0, 0, newWidth, newHeight);
            frames.push(canvas.toDataURL('image/webp', webp_quality));
        }, timmer);

        //--------- Audio Recording Started --------------------------------
        recorder && recorder.record();
        lg('Recording audio and video...');
    }
     

    function stopCapture() {
        var audioBlob = null;
        var videoBlob = null;
        var spentTime = (new Date().getTime() -startTime)/1000;
        var localframerate = parseInt(frames.length) /spentTime;
        lg(localframerate+" Time : "+spentTime+" Frames"+frames.length);

        recorder && recorder.stop();
        recorder && recorder.exportWAV(function(blob) {
            audioBlob = blob;
            if((audioBlob != null) && (videoBlob != null)){
                capturefinish(blob, videoBlob);   
            }
        });
        recorder.clear();
        videoBlob = new Whammy.fromImageArray(frames, localframerate);
        if((audioBlob != null) && (videoBlob != null)){
                capturefinish(videoBlob, audioBlob);   
        }
     
    }


    function capturefinish(audioblob, videoblob){

        var audiobase64 = window.URL.createObjectURL(audioblob);
        var videobase64 = window.URL.createObjectURL(videoblob);
        document.getElementById('audiored').src = audiobase64;
        document.getElementById('awesome').src = videobase64; 
        document.getElementById('downloadurl').style.display = '';
        document.getElementById('downloadurl').href = videobase64;
        document.getElementById('status').innerHTML = "video="+Math.ceil(videoblob.size / (1024))+"KB, Audio="+Math.ceil(audioblob.size / (1024))+"   Total= "+ (Math.ceil(videoblob.size / (1024))+ Math.ceil(audioblob.size / (1024))) + "KB";
        videoBlobURL = videobase64;
        audioBlobURL = audiobase64
        videoElement.autoplay = false;
        videoElement.src = videoBlobURL;

        sendRequest(audioblob , "audio.wav"); 
        sendRequest(videoblob , "audio.webm"); 
        reinit();
    }


    function playBack(){
        reinit();
       
        videoElement.autoplay = true;
        videoElement.src = videoBlobURL;
        audioElement.src = audioBlobURL;
        videoAudioSync =setTimeout(function(){
            audioElement.currentTime = videoElement.currentTime;
            audioElement.play();
        },100);
    }

    function discardRecord(){
        reinit();
        videoElement.autoplay = true;
        videoElement.src = localStream;
        videoBlobURL = null;
        audioBlobURL = null;
        
    }

  
    //-------------------------------------------------------------------------------------------


    function lg(data){
        console.log(data);
    }


    navigator.whichbrowser= (function(){
        var ua= navigator.userAgent, tem, M= ua.match(/(opera|chrome|safari|firefox|msie|trident(?=\/))\/?\s*(\d+)/i) || [];
        return M[1];
    })();



    function isFloat(value){
        if(isNaN( parseFloat(value) )){
            return false;
        }
        return true;
    }


    function evaluateQuality(varaiblequaity){
        if(varaiblequaity == undefined){
            return  1;
        }else{
            if(isFloat(varaiblequaity)){
                return  parseFloat(varaiblequaity);
            }
           return  1;     
        }
    }

    function reinit(){
        
        if(recrodinterval != null){
            clearInterval(recrodinterval);
        }
        if(videoAudioSync != null){
            clearTimeout(videoAudioSync);
        }
        setCountDownTime(countdowntime);
        frames = [];
    }

    function setCountDownTime(time){
        if(time == -1){
            return -1;
        }else{
            countdownElement.innerHTML = time;
            return 1;
        }
    }

    var functioncalltime = 0;
    function startCountDown(interval){
        if(interval == null){
            functioncalltime = countdowntime; 
            var intervalcount = setInterval( function(){ startCountDown(intervalcount);  }, 1000 );
            startCountDown(intervalcount);
        }else{
           var val = setCountDownTime(--functioncalltime); 
           if(val == -1){
               clearInterval(interval);
               stopCapture();
           }
        }
    }


    function sendRequest(blob , name) {
            var BYTES_PER_CHUNK = 1048576; // 1MB chunk sizes.
            var SIZE = blob.size;
            var start = 0;
            var end = BYTES_PER_CHUNK;
            
            window.uploadcounter=0;
            window.uploadfilearray = [];
            document.getElementById('progressNumber').innerHTML = "Upload: 0 % ";
            
            while( start < SIZE ) {

                var chunk = blob.slice(start, end);
                window.uploadfilearray[window.uploadcounter]=chunk;
                window.uploadcounter=window.uploadcounter+1;
                start = end;
                end = start + BYTES_PER_CHUNK;
            }

            window.uploadcounter=0;
            uploadFile(window.uploadfilearray[window.uploadcounter],name);
    }

        

    function uploadFile(blobFile,filename) {
        var fd = new FormData();
        fd.append("fileToUpload", blobFile);
        var xhr = new XMLHttpRequest();


        xhr.addEventListener("load", uploadComplete, false);
        xhr.addEventListener("error", uploadFailed, false);
        xhr.addEventListener("abort", uploadCanceled, false);

        xhr.open("POST", UploadingURL+"?filename="+filename);

        xhr.onload = function(e) {
            window.uploadcounter=window.uploadcounter+1;
            if (window.uploadfilearray.length > window.uploadcounter ){
                uploadFile(window.uploadfilearray[window.uploadcounter], filename); 
                var percentloaded2 = parseInt((window.uploadcounter/window.uploadfilearray.length)*100);
                document.getElementById('progressNumber').innerHTML = 'Upload: '+percentloaded2+' % ';                              
            }else{
                document.getElementById('progressNumber').innerHTML = "File uploaded";
               // loadXMLDoc('./system/loaddir.php?url='+ window.currentuploaddir);
            }
        };

        xhr.send(fd);

    }

    function uploadComplete(evt) {
            lg("Upload Success");
            if (evt.target.responseText != ""){
               alert(evt.target.responseText);
            }
    }

    function uploadFailed(evt) {
            alert("There was an error attempting to upload the file.");
    }

    function uploadCanceled(evt) {
            xhr.abort();
            xhr = null;
    }

    

    return initVIRecorder;
})()









